File: app/src/main/java/com/ivanovsky/passnotes/data/repository/file/FileSystemProvider.java
package com.ivanovsky.passnotes.data.repository.file;

import androidx.annotation.NonNull;

import com.ivanovsky.passnotes.data.entity.FileDescriptor;
import com.ivanovsky.passnotes.data.entity.OperationResult;

import java.io.InputStream;
import java.io.OutputStream;
import java.util.List;

public interface FileSystemProvider {

    @NonNull
    FileSystemAuthenticator getAuthenticator();

    @NonNull
    FileSystemSyncProcessor getSyncProcessor();

    @NonNull
    OperationResult<List<FileDescriptor>> listFiles(@NonNull FileDescriptor dir);

    @NonNull
    OperationResult<FileDescriptor> getParent(@NonNull FileDescriptor file);

    @NonNull
    OperationResult<FileDescriptor> getRootFile();

    @NonNull
    OperationResult<InputStream> openFileForRead(@NonNull FileDescriptor file,
                                                 @NonNull OnConflictStrategy onConflictStrategy,
                                                 @NonNull FSOptions options);

    @NonNull
    OperationResult<OutputStream> openFileForWrite(@NonNull FileDescriptor file,
                                                   @NonNull OnConflictStrategy onConflictStrategy,
                                                   @NonNull FSOptions options);

    @NonNull
    OperationResult<Boolean> exists(@NonNull FileDescriptor file);

    @NonNull
    OperationResult<FileDescriptor> getFile(@NonNull String path,
                                            @NonNull FSOptions options);
}

File: app/src/main/kotlin/com/ivanovsky/passnotes/data/repository/file/FileSystemProvider.kt
package com.ivanovsky.passnotes.data.repository.file

import com.ivanovsky.passnotes.data.entity.FileDescriptor
import com.ivanovsky.passnotes.data.entity.OperationResult
import java.io.InputStream
import java.io.OutputStream

interface FileSystemProvider {

    val authenticator: FileSystemAuthenticator
    val syncProcessor: FileSystemSyncProcessor

    fun listFiles(dir: FileDescriptor): OperationResult<List<FileDescriptor>>
    fun getParent(file: FileDescriptor): OperationResult<FileDescriptor>
    fun getRootFile(): OperationResult<FileDescriptor>
    fun exists(file: FileDescriptor): OperationResult<Boolean>
    fun getFile(path: String, options: FSOptions): OperationResult<FileDescriptor>

    fun openFileForRead(
        file: FileDescriptor,
        onConflictStrategy: OnConflictStrategy,
        options: FSOptions
    ): OperationResult<InputStream>

    fun openFileForWrite(
        file: FileDescriptor,
        onConflictStrategy: OnConflictStrategy,
        options: FSOptions
    ): OperationResult<OutputStream>
}